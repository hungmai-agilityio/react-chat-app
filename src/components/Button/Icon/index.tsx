import { clsx } from 'clsx';

// FontAwesome
import { IconDefinition } from '@fortawesome/fontawesome-svg-core';
import { FontAwesomeIcon } from '@fortawesome/react-fontawesome';

interface IconProps {
  icon: IconDefinition;
  onClick?: () => void;
  styles?: string;
  isHover?: boolean;
}

const Icon = ({ icon, styles, isHover = true, onClick }: IconProps) => {
  return (
    <div
      className={clsx(
        'flex items-center justify-center cursor-pointer text-gray-400 h-10 w-10 rounded-full',
        { 'hover:text-dark hover:bg-tertiary': isHover },
        styles
      )}
      onClick={onClick}
    >
      <FontAwesomeIcon icon={icon} />
    </div>
  );
};

export default Icon;
