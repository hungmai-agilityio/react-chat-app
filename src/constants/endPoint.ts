export const URL = {
  HOME: '',
  SIGN_IN: '/sign-in',
  SIGN_UP: '/sign-up'
}

export const END_POINT = {
  USER: '/users',
  PROFILE: '/profiles',
  CHAT: 'chats',
  MESSAGE: 'messages'
}