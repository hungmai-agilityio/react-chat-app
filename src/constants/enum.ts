enum SIZE {
  MINI = 'extra-small',
  SMALL = 'small',
  MEDIUM = 'medium',
  LARGE = 'large'
}

enum TYPE {
  PRIMARY = 'primary',
  SECOND = 'second',
  TERTIARY = 'tertiary',
  FILE = 'file',
  TEXT = 'text',
  PASSWORD = 'password'
}

enum POSITION {
  TOP_RIGHT = 'top-right',
  TOP_LEFT = 'top-left',
  BOT_RIGHT = 'bottom-right',
  BOT_LEFT = 'bottom-left'
}

export { SIZE, TYPE, POSITION };
