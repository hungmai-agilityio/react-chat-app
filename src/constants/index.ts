export * from '@/constants/enum';
export * from '@/constants/list';
export * from '@/constants/endPoint';
export * from '@/constants/regex';
export * from '@/constants/message'
