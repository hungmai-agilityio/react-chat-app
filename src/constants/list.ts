import { ITab } from '@/components/Tabs/Tab';
import { faEdit, faTrash } from '@fortawesome/free-solid-svg-icons';

export const USER_OPTIONS = [{ label: 'Profile', value: 'profile' }, { label: 'Log out', value: 'logout' }];

export const MESSAGE_OPTIONS = [
  { label: 'Edit', value: 'editMess', icon: faEdit },
  { label: 'Delete', value: 'deleteMess', icon: faTrash }
];

export const INFO_OPTIONS = [{ label: 'Chat info', value: 'infoChat' }];

export const LIST_TAB_USERS: Omit<ITab, 'index' | 'onClick'>[] = [
  {
    title: 'Chat',
    value: '1'
  },
  {
    title: 'Users',
    value: '2'
  }
];