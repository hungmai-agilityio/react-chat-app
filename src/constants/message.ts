export const MESSAGE_VALID = {
  INPUT_REQUIRED: '{field} is required',
  EMAIL_ERROR: 'Email is not valid',
  EMAIL_EXIST: 'Email has been registered, please try again!',
  PASSWORD_ERROR:
    'Minimum 6 characters, one uppercase, one lowercase, one number',
  CONFIRM_ERROR: 'Password confirmation do not match password',
};

export const MESSAGE_API = {
  SIGN_UP_ERROR: 'An error occurred. Please try again',
  SIGN_IN_ERROR: 'email or password is incorrect',
  UPDATE_PROFILE_SUCCESS: 'Update success',
  UPDATE_PROFILE_ERROR: 'Update error',
};