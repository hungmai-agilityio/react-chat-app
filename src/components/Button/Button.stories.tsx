import { Meta, StoryObj } from '@storybook/react';

// Components
import Button from '@/components/Button';

// Constants
import { SIZE, TYPE } from '@/constants';

// FontAwesome
import { faPen } from '@fortawesome/free-solid-svg-icons';

const meta: Meta<typeof Button> = {
  component: Button,
  tags: ['autodocs'],
  argTypes: {
    name: {
      description: 'Name button',
    },
    size: {
      description: 'Set size optional for Button'
    },
    variant: {
      description: 'Set type optional for Button'
    },
    onClick: {
      description: 'Handle the event when clicking on the button'
    },
    styles: {
      description: 'Add custom css'
    },
    isDisabled: {
      description: 'Set disabled button'
    },
    iconLeft: {
      description: 'Render icon left with FontAwesome'
    },
    iconRight: {
      description: 'Render icon right with FontAwesome'
    }
  }
};

export default meta;

type Story = StoryObj<typeof Button>;

export const Primary: Story = {
  args: {
    name: 'Submit',
    size: SIZE.MEDIUM,
    variant: TYPE.PRIMARY,
    onClick: () => alert('submit primary button')
  }
};

export const Secondary: Story = {
  args: {
    name: 'Submit',
    size: SIZE.MEDIUM,
    variant: TYPE.SECOND,
    onClick: () => alert('submit secondary button')
  }
};

export const Tertiary: Story = {
  args: {
    name: 'Submit',
    size: SIZE.SMALL,
    variant: TYPE.TERTIARY,
    iconRight: faPen,
    onClick: () => alert('submit Tertiary button')
  }
};

export const ButtonLarge: Story = {
  args: {
    name: 'Submit',
    size: SIZE.LARGE,
    variant: TYPE.TERTIARY,
    iconLeft: faPen,
    onClick: () => alert('submit secondary button')
  }
};

export const ButtonDisabled: Story = {
  args: {
    name: 'Send',
    size: SIZE.MINI,
    isDisabled: true,
    onClick: () => alert('Can not submit')
  }
};
