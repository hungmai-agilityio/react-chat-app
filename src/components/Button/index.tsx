import { clsx } from 'clsx';
import { memo } from 'react';

// Constants
import { SIZE, TYPE } from '@/constants';

// FontAwesome
import { IconDefinition } from '@fortawesome/fontawesome-svg-core';
import { FontAwesomeIcon } from '@fortawesome/react-fontawesome';

interface ButtonProps {
  name: string;
  size?: SIZE.MINI | SIZE.SMALL | SIZE.MEDIUM | SIZE.LARGE;
  variant?: TYPE.PRIMARY | TYPE.SECOND | TYPE.TERTIARY;
  onClick: (e: React.FormEvent<HTMLButtonElement>) => void;
  styles?: string;
  isDisabled?: boolean;
  iconLeft?: IconDefinition;
  iconRight?: IconDefinition;
}

const Button = ({
  name,
  size,
  variant,
  onClick,
  styles,
  isDisabled,
  iconLeft,
  iconRight
}: ButtonProps) => {
  const sizeClasses = {
    [SIZE.MINI]: 'w-24 h-9',
    [SIZE.SMALL]: 'w-32 h-10',
    [SIZE.MEDIUM]: 'w-40 h-9',
    [SIZE.LARGE]: 'w-full h-10'
  };

  const typeClasses = {
    [TYPE.PRIMARY]: 'bg-primary text-white',
    [TYPE.SECOND]: 'bg-tertiary text-primary',
    [TYPE.TERTIARY]: 'hover:bg-tertiary text-primary'
  };

  const sizeClass = sizeClasses[size || SIZE.SMALL];
  const typeClass = typeClasses[variant || TYPE.PRIMARY];

  return (
    <button
      className={clsx(
        'font-semibold hover:brightness-90',
        isDisabled &&
          'bg-tertiary cursor-not-allowed border hover:filter-none text-stone-500',
        sizeClass,
        typeClass,
        styles
      )}
      onClick={onClick}
      disabled={isDisabled}
    >
      {iconLeft && <FontAwesomeIcon icon={iconLeft} className="mr-2" />}
      {name}
      {iconRight && <FontAwesomeIcon icon={iconRight} className="ml-2" />}
    </button>
  );
};

const MemoizedButton = memo(Button);
export default MemoizedButton;
