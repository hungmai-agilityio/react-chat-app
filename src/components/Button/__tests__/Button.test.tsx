import { fireEvent, render, screen } from '@testing-library/react';
import '@testing-library/jest-dom';

// Components
import Button from '@/components/Button';

// Constants
import { SIZE, TYPE } from '@/constants';

// FontAwesome
import { faAdd, faEdit } from '@fortawesome/free-solid-svg-icons';

describe('Button Component', () => {
  const onClickMock = jest.fn();

  it('renders default Button', () => {
    const { getByText } = render(
      <Button name="Click me" onClick={onClickMock} />
    );

    const buttonElement = getByText('Click me');
    expect(buttonElement).toBeInTheDocument();
  });

  it('renders custom Button', () => {
    const { getByText } = render(
      <Button
        name="Click me"
        onClick={onClickMock}
        size={SIZE.MEDIUM}
        variant={TYPE.SECOND}
      />
    );

    const buttonElement = getByText('Click me');
    expect(buttonElement).toBeInTheDocument();
  });

  it('renders custom with icon', () => {
    const { getByText } = render(
      <Button
        name="Click me"
        onClick={onClickMock}
        size={SIZE.MEDIUM}
        variant={TYPE.SECOND}
        iconLeft={faEdit}
        iconRight={faAdd}
      />
    );

    const buttonElement = getByText('Click me');
    expect(buttonElement).toBeInTheDocument();
  });

  it('renders disabled Button', () => {
    render(
      <Button
        name="Click me"
        onClick={onClickMock}
        size={SIZE.MEDIUM}
        variant={TYPE.SECOND}
        isDisabled
      />
    );

    const buttonElement = screen.getByRole('button');
    expect(buttonElement).toBeDisabled();
  });

  it('calls onClick handler when clicked', () => {
    const { getByText } = render(
      <Button name="Click me" onClick={onClickMock} />
    );

    const buttonElement = getByText('Click me');
    fireEvent.click(buttonElement);

    expect(onClickMock).toHaveBeenCalled();
  });

  it('should be render match to snapshot', () => {
    const { container } = render(
      <Button name="Click me" onClick={onClickMock} size={SIZE.MEDIUM} />
    );

    expect(container).toMatchSnapshot();
  });
});
