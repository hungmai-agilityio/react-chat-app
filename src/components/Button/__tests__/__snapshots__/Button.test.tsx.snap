// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Button Component should be render match to snapshot 1`] = `
<div>
  <button
    class="font-semibold hover:brightness-90 w-40 h-9 bg-primary text-white"
  >
    Click me
  </button>
</div>
`;
